File: pkgs/ast/src/markup.ts
/** What `reduce` emits. Not the IR: nothing here has been decided yet, only narrowed. */
export type MarkupNode =
	| { k: 'text'; v: string }
	| { k: 'expr'; src: string }
	| { k: 'html'; src: string }
	| { k: 'element'; name: string; attrs: MarkupAttr[]; body: MarkupNode[] }
	// consequent/alternate rather than then/else: an object carrying a `then` is thenable, so
	// awaiting one anywhere would call it. The names are Svelte's own for the same fields.
	| { k: 'if'; test: string; consequent: MarkupNode[]; alternate: MarkupNode[] | null }
	| {
			k: 'each';
			source: string;
			item: string | null;
			index: string | null;
			key: string | null;
			body: MarkupNode[];
			fallback: MarkupNode[] | null;
	  }
	| { k: 'component'; name: string; props: MarkupAttr[]; body: MarkupNode[] }
	| { k: 'unsupported'; type: string; src: string };

export type MarkupAttr =
	| { k: 'attr'; name: string; value: true | MarkupNode[] }
	| { k: 'unsupported'; type: string; src: string };

/** One component file. `imports` is every import it declares, unfiltered. */
export interface Module {
	markup: MarkupNode[];
	imports: Record<string, string>;
}

/** An entry and everything reachable from it, keyed by an id derived from the file path. */
export interface Bundle {
	entry: string;
	components: Record<string, Module>;
}

File: crates/lowering/src/markup.rs
//! The shape `pkgs/ast` emits. It has decided nothing: every expression is still source text.

use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(tag = "k", rename_all = "lowercase")]
pub enum Node {
	Text {
		v: String,
	},
	Expr {
		src: String,
	},
	Html {
		src: String,
	},
	Element {
		name: String,
		attrs: Vec<Attr>,
		body: Vec<Node>,
	},
	If {
		test: String,
		consequent: Vec<Node>,
		alternate: Option<Vec<Node>>,
	},
	Each {
		source: String,
		item: Option<String>,
		index: Option<String>,
		key: Option<String>,
		body: Vec<Node>,
		fallback: Option<Vec<Node>>,
	},
	Component {
		name: String,
		props: Vec<Attr>,
		body: Vec<Node>,
	},
	Unsupported {
		#[serde(rename = "type")]
		kind: String,
		src: String,
	},
}

#[derive(Debug, Deserialize)]
#[serde(tag = "k", rename_all = "lowercase")]
pub enum Attr {
	Attr {
		name: String,
		value: AttrValue,
	},
	Unsupported {
		#[serde(rename = "type")]
		kind: String,
		src: String,
	},
}

#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub enum AttrValue {
	Present(bool),
	Parts(Vec<Node>),
}

/// One component file, with its imports already resolved to bundle ids by whoever read them.
#[derive(Debug, Deserialize)]
pub struct Module {
	pub markup: Vec<Node>,
	pub imports: std::collections::BTreeMap<String, String>,
}

/// An entry and everything reachable from it.
#[derive(Debug, Deserialize)]
pub struct Bundle {
	pub entry: String,
	pub components: std::collections::BTreeMap<String, Module>,
}
